server/src/main/java/org/elasticsearch/common/util/concurrent/AtomicArray.java

```diff
@@ -98,20 +100,20 @@ public List<E> asList() {
                         list.add(e);
                     }
                 }
-                nonNullList = list;
+                res = list;
             }
+            nonNullList = res;
         }
-        return nonNullList;
+        return res;
     }
 
     /**
      * Copies the content of the underlying atomic array to a normal one.
      */
-    public E[] toArray(E[] a) {
```

Comment:
All usage of this thing consumed an array sized via `AtomicArray#length` might as well create the array in here and fill it by looping which should avoid `calloc` (at least it does for most JDK collections where the logic is kinda the same).

Reply:
All specifications of `toArray` in the JCL have the argument as `Class<E[]>`, doing different here is confusing

Reply:
I think that's less efficient though and blows up the code size more, even Java's own thing just does this to then get the class from a zero sized array:

```
    default <T> T[] toArray(IntFunction<T[]> generator) {
        return toArray(generator.apply(0));
    }
```
It makes sense because it wouldn't be able to eliminate the nulling out of the array inside the lambda `generator`.

=> So either we'd take worse performance here or weirdly complicated code if we went for that API?

Reply:
But here you can use the generator directly:

    E[] a = generator.apply(array.length())

Reply:
Jup but my point was, the JDK doesn't do that either for performance reasons and instead uses the `0` length array step. If you use the generator directly the JVM probably won't be able to optimise away the nulling out of the array :)

Reply:
Are you sure? The function is a direct ref to an array constructor, the JVM can probably inline the function and then do the reacheablility analysis to determine it doesnt need to null out each item before it gets filled in directly below.

That said, we're arguing over specific details, and the old code is definitely doing the nullability check. I think it will be able to elide the null assignments, but I'm more in favour of the consistency of the API with already-established patterns, unless this is a known specific bottleneck that is improved by creating the array directly.

Reply:
Yea pretty sure.
The crux is this:

> can probably inline the function and then do the reacheablility analysis to determine

It needs the inlining to do the subsequent optimization. The inlining won't happen though because there's like 10 different lambdas that can be passed into the method (~10 callsites in our codebase) -> megamorphic call-site -> no inlining pretty much

Reply:
I'm still in favour of consistency, given that the current code does null assignment (and so we're not losing anything), unless this is a specific bottleneck that needs improvement in this way

Reply:
That would also be ok - pass in a zero-length array, and resize it if needed (same as other implementations of `toArray(E[])`